src/managed/Microsoft.Extensions.DependencyModel/UnifiedJsonReader.Utf8JsonReader.cs

```diff
@@ -0,0 +1,166 @@
+            {
+                throw CreateUnexpectedException(ref _reader, "}");
+            }
+        }
+
+        public string[] ReadStringArray()
+        {
+            _reader.Read();
+            if (_reader.TokenType != JsonTokenType.StartArray)
+            {
+                throw CreateUnexpectedException(ref _reader, "[");
+            }
+
+            var items = new List<string>();
+
+            while (_reader.Read() && _reader.TokenType == JsonTokenType.String)
+            {
+                items.Add(GetStringValue());
+            }
+
+            if (_reader.TokenType != JsonTokenType.EndArray)
+            {
+                throw CreateUnexpectedException(ref _reader, "]");
+            }
+
+            return items.ToArray();
+        }
+
+        public void Skip()
+        {
+            if (IsTokenTypeProperty())
+            {
+                _reader.Read();
+            }
+
+            if (_reader.TokenType == JsonTokenType.StartObject || _reader.TokenType == JsonTokenType.StartArray)
+            {
+                int depth = _reader.CurrentDepth;
+                while (_reader.Read() && depth <= _reader.CurrentDepth)
+                {
+                }
+            }
+        }
+
+        public string ReadAsString()
+        {
+            Debug.Assert(IsTokenTypeProperty());
+            _reader.Read();
+            if (_reader.TokenType == JsonTokenType.Null)
+            {
+                return null;
+            }
+            if (_reader.TokenType != JsonTokenType.String)
+            {
+                throw CreateUnexpectedException(ref _reader, "a JSON string token");
+            }
+            Debug.Assert(!IsTokenTypeProperty());
+            return GetStringValue();
+        }
+
+        public bool? ReadAsNullableBoolean()
+        {
+            Debug.Assert(IsTokenTypeProperty());
+            _reader.Read();
+            if (_reader.TokenType == JsonTokenType.Null)
+            {
+                return null;
+            }
+            if (_reader.TokenType != JsonTokenType.True && _reader.TokenType != JsonTokenType.False)
+            {
+                throw CreateUnexpectedException(ref _reader, "a JSON true or false literal token");
+            }
+            return _reader.GetBoolean();
+        }
+
+        public bool ReadAsBoolean(bool defaultValue)
+        {
+            Debug.Assert(IsTokenTypeProperty());
+            _reader.Read();
+            if (_reader.TokenType == JsonTokenType.Null)
+            {
+                return defaultValue;
+            }
+            if (_reader.TokenType != JsonTokenType.True && _reader.TokenType != JsonTokenType.False)
+            {
+                throw CreateUnexpectedException(ref _reader, "a JSON true or false literal token");
+            }
+            return _reader.GetBoolean();
+        }
+
+        private static Exception CreateUnexpectedException(ref Utf8JsonReader reader, string expected)
```

Comment:
(minor) It'd be nice to have to not have this method duplicated.

Reply:
To keep the current custom exception message, it has to be. JsonTextReader returns the path.
I can extract out the creation of the exception message, but that's about it.

Reply:
OK, this makes sense. Thanks!

---
In reply to: [251344200](https://github.com/dotnet/core-setup/pull/5009#discussion_r251344200) [](ancestors = 251344200)